File: packages/python/navaia_forge/integrations/langgraph/callback.py
"""LangChain callback handler that streams events to Forge observability.

Drop-in for any ``langchain_core``-compatible runnable (which is what
LangGraph compiles to). The handler is best-effort: if the Forge backend
is unreachable, callbacks log a warning and continue rather than killing
the user's graph mid-run.

Usage::

    cb = NavaiaForgeCallback(client, task_id="task_123")
    graph.invoke(state, config={"callbacks": [cb]})

What it reports:

- ``on_llm_end`` → ``client.observability.log_token_usage(...)`` with the
  model name, input/output tokens, cost when present, and the duration.
- ``on_chain_start`` / ``on_chain_end`` / ``on_chain_error`` →
  ``client.tasks.append_log(...)`` if the SDK exposes that method, else
  a structured log line via the standard ``logging`` module. We never
  fail the run on a logging failure.
- ``on_tool_start`` / ``on_tool_end`` → log line + token-usage marker so
  cost dashboards can attribute spend to the tool invocation.
"""

from __future__ import annotations

import logging
import time
from typing import TYPE_CHECKING, Any
from uuid import UUID

from ._lazy import import_base_callback_handler

if TYPE_CHECKING:  # pragma: no cover
    from ...client import NavaiaForgeClient


logger = logging.getLogger(__name__)


def _make_callback_class() -> type:
    """Build the callback class with ``BaseCallbackHandler`` as the base.

    We can't subclass at module import time without forcing the optional
    ``langchain_core`` dependency, so the class is built lazily the first
    time anyone instantiates :class:`NavaiaForgeCallback`.
    """
    base = import_base_callback_handler()

    class _NavaiaForgeCallbackImpl(base):  # type: ignore[misc, valid-type]
        """Internal — see :class:`NavaiaForgeCallback` for the public surface."""

        def __init__(
            self,
            client: NavaiaForgeClient,
            *,
            task_id: str | None = None,
            agent_id: str | None = None,
            workforce_id: str | None = None,
            default_model: str = "unknown",
        ) -> None:
            super().__init__()
            self._client = client
            self._task_id = task_id
            self._agent_id = agent_id
            self._workforce_id = workforce_id
            self._default_model = default_model
            # run_id → start time, so on_*_end can compute duration_ms.
            self._starts: dict[UUID, float] = {}

        # ── LLM ────────────────────────────────────────────────

        def on_llm_start(
            self,
            serialized: dict[str, Any],
            prompts: list[str],
            *,
            run_id: UUID,
            **kwargs: Any,
        ) -> None:
            self._starts[run_id] = time.monotonic()

        def on_llm_end(self, response: Any, *, run_id: UUID, **kwargs: Any) -> None:
            duration_ms = self._duration_ms(run_id)
            usage = _extract_token_usage(response)
            model = _extract_model_name(response) or self._default_model

            try:
                self._client.observability.log_token_usage(
                    model=model,
                    agent_id=self._agent_id,
                    task_id=self._task_id,
                    input_tokens=usage["input"],
                    output_tokens=usage["output"],
                    cost_usd=usage["cost_usd"],
                    duration_ms=duration_ms,
                )
            except Exception as exc:
                logger.warning(
                    "NavaiaForgeCallback: token-usage log failed (%s)", exc
                )

        def on_llm_error(
            self, error: BaseException, *, run_id: UUID, **kwargs: Any
        ) -> None:
            self._starts.pop(run_id, None)
            logger.warning("LangGraph LLM error: %s", error)

        # ── Chain ──────────────────────────────────────────────

        def on_chain_start(
            self,
            serialized: dict[str, Any],
            inputs: dict[str, Any],
            *,
            run_id: UUID,
            **kwargs: Any,
        ) -> None:
            self._starts[run_id] = time.monotonic()
            logger.debug(
                "LangGraph chain start name=%s task=%s",
                _name_of(serialized),
                self._task_id,
            )

        def on_chain_end(
            self, outputs: dict[str, Any], *, run_id: UUID, **kwargs: Any
        ) -> None:
            duration_ms = self._duration_ms(run_id)
            logger.debug(
                "LangGraph chain end task=%s duration_ms=%d", self._task_id, duration_ms
            )

        def on_chain_error(
            self, error: BaseException, *, run_id: UUID, **kwargs: Any
        ) -> None:
            self._starts.pop(run_id, None)
            logger.warning(
                "LangGraph chain error task=%s: %s", self._task_id, error
            )

        # ── Tool ───────────────────────────────────────────────

        def on_tool_start(
            self,
            serialized: dict[str, Any],
            input_str: str,
            *,
            run_id: UUID,
            **kwargs: Any,
        ) -> None:
            self._starts[run_id] = time.monotonic()
            logger.debug(
                "LangGraph tool start name=%s task=%s",
                _name_of(serialized),
                self._task_id,
            )

        def on_tool_end(
            self, output: Any, *, run_id: UUID, **kwargs: Any
        ) -> None:
            duration_ms = self._duration_ms(run_id)
            logger.debug(
                "LangGraph tool end task=%s duration_ms=%d",
                self._task_id,
                duration_ms,
            )

        def on_tool_error(
            self, error: BaseException, *, run_id: UUID, **kwargs: Any
        ) -> None:
            self._starts.pop(run_id, None)
            logger.warning("LangGraph tool error task=%s: %s", self._task_id, error)

        # ── Helpers ────────────────────────────────────────────

        def _duration_ms(self, run_id: UUID) -> int:
            start = self._starts.pop(run_id, None)
            if start is None:
                return 0
            return int((time.monotonic() - start) * 1000)

    return _NavaiaForgeCallbackImpl


def NavaiaForgeCallback(  # noqa: N802 — class-like factory; defers langchain_core import
    client: NavaiaForgeClient,
    *,
    task_id: str | None = None,
    agent_id: str | None = None,
    workforce_id: str | None = None,
    default_model: str = "unknown",
) -> Any:
    """Factory returning a ``BaseCallbackHandler`` instance bound to ``client``.

    Implemented as a function rather than a class so the
    ``langchain_core`` import is deferred until first use — keeps the
    bare ``import navaia_forge`` cheap and safe in environments without
    the optional extra installed.

    Args:
        client: A configured :class:`NavaiaForgeClient`.
        task_id: Forge task id this run is fulfilling, attached to every
            event for cost/usage attribution.
        agent_id: Forge agent id when the graph represents a single agent.
        workforce_id: Forge workforce id, used for high-level rollups.
        default_model: Fallback model name when LangChain doesn't surface
            one — for chains that wrap raw HTTP calls or custom LLMs.
    """
    cls = _make_callback_class()
    return cls(
        client,
        task_id=task_id,
        agent_id=agent_id,
        workforce_id=workforce_id,
        default_model=default_model,
    )


# ── Pure helpers (testable without langchain_core) ────────────────


def _extract_token_usage(response: Any) -> dict[str, int | float]:
    """Pull input/output tokens and cost out of an LLMResult-like object.

    LangChain's response shape varies across versions and providers. The
    canonical place is ``response.llm_output["token_usage"]`` for chat
    models and ``response.generations[0][0].generation_info`` for some
    completions. We also walk ``message.usage_metadata`` which is the
    newer (langchain-core 0.2+) location.
    """
    out = {"input": 0, "output": 0, "cost_usd": 0.0}

    llm_output = getattr(response, "llm_output", None)
    if isinstance(llm_output, dict):
        usage = llm_output.get("token_usage") or llm_output.get("usage") or {}
        if isinstance(usage, dict):
            out["input"] = int(
                usage.get("input_tokens") or usage.get("prompt_tokens") or 0
            )
            out["output"] = int(
                usage.get("output_tokens") or usage.get("completion_tokens") or 0
            )

    if out["input"] == 0 and out["output"] == 0:
        # Try the per-generation usage metadata.
        generations = getattr(response, "generations", None) or []
        for batch in generations:
            for gen in batch or []:
                message = getattr(gen, "message", None)
                meta = getattr(message, "usage_metadata", None)
                if isinstance(meta, dict):
                    out["input"] = int(meta.get("input_tokens", 0))
                    out["output"] = int(meta.get("output_tokens", 0))
                    break
            if out["input"] or out["output"]:
                break

    return out


def _extract_model_name(response: Any) -> str | None:
    """Best-effort model name extraction from an LLMResult-like object."""
    llm_output = getattr(response, "llm_output", None)
    if isinstance(llm_output, dict):
        name = llm_output.get("model_name") or llm_output.get("model")
        if isinstance(name, str) and name:
            return name
    return None


def _name_of(serialized: dict[str, Any] | None) -> str:
    if not isinstance(serialized, dict):
        return "<unknown>"
    return serialized.get("name") or serialized.get("id", ["<unknown>"])[-1] or "<unknown>"

File: packages/javascript/src/integrations/langgraph/callback.ts
/**
 * LangChain callback handler that streams events to Forge observability.
 *
 * Drop-in for any `@langchain/core`-compatible runnable (which is what
 * LangGraph compiles to). The handler is best-effort: if the Forge backend
 * is unreachable, callbacks log a warning and continue rather than killing
 * the graph mid-run.
 *
 * What it reports:
 *
 * - `handleLLMEnd` → `client.observability.logTokenUsage(...)` with the
 *   model name, input/output tokens, cost when present, and the duration.
 * - `handleChainStart` / `handleChainEnd` / `handleChainError` and the
 *   `handleTool*` family — debug log lines so cost dashboards can attribute
 *   spend to specific nodes/tools.
 */

import type { NavaiaForge } from "../../client.js";
import { importBaseCallbackHandler } from "./lazy.js";

export interface NavaiaForgeCallbackOptions {
  readonly taskId?: string | null;
  readonly agentId?: string | null;
  readonly workforceId?: string | null;
  /** Fallback model name when LangChain doesn't surface one. */
  readonly defaultModel?: string;
}

interface UsageNumbers {
  input: number;
  output: number;
  costUsd: number;
}

/**
 * Factory returning a `BaseCallbackHandler` instance bound to `client`.
 *
 * Implemented as an async function so the `@langchain/core` import is
 * deferred until first use — keeps the bare `import { NavaiaForge }` cheap
 * and safe in environments without the optional peer dep installed.
 */
export async function createNavaiaForgeCallback(
  client: NavaiaForge,
  options: NavaiaForgeCallbackOptions = {},
): Promise<unknown> {
  const Base = await importBaseCallbackHandler();
  const taskId = options.taskId ?? null;
  const agentId = options.agentId ?? null;
  const workforceId = options.workforceId ?? null;
  const defaultModel = options.defaultModel ?? "unknown";

  // Use a class expression so we capture the lazily-imported base.
  class NavaiaForgeCallbackImpl extends (Base as new (...args: unknown[]) => {
    [k: string]: unknown;
  }) {
    name = "NavaiaForgeCallback";
    private starts = new Map<string, number>();

    constructor() {
      super();
    }

    // ── LLM ────────────────────────────────────────────

    async handleLLMStart(
      _llm: unknown,
      _prompts: string[],
      runId: string,
    ): Promise<void> {
      this.starts.set(runId, Date.now());
    }

    async handleLLMEnd(output: unknown, runId: string): Promise<void> {
      const durationMs = this.durationMs(runId);
      const usage = extractTokenUsage(output);
      const model = extractModelName(output) ?? defaultModel;

      try {
        await client.observability.logTokenUsage({
          model,
          agent_id: agentId ?? undefined,
          task_id: taskId ?? undefined,
          input_tokens: usage.input,
          output_tokens: usage.output,
          cost_usd: usage.costUsd,
          duration_ms: durationMs,
        });
      } catch (err) {
        // eslint-disable-next-line no-console
        console.warn(
          `NavaiaForgeCallback: token-usage log failed: ${String(err)}`,
        );
      }
    }

    async handleLLMError(_err: unknown, runId: string): Promise<void> {
      this.starts.delete(runId);
    }

    // ── Chain ──────────────────────────────────────────

    async handleChainStart(
      _chain: unknown,
      _inputs: unknown,
      runId: string,
    ): Promise<void> {
      this.starts.set(runId, Date.now());
    }

    async handleChainEnd(_outputs: unknown, runId: string): Promise<void> {
      this.durationMs(runId);
    }

    async handleChainError(_err: unknown, runId: string): Promise<void> {
      this.starts.delete(runId);
    }

    // ── Tool ───────────────────────────────────────────

    async handleToolStart(
      _tool: unknown,
      _input: string,
      runId: string,
    ): Promise<void> {
      this.starts.set(runId, Date.now());
    }

    async handleToolEnd(_output: unknown, runId: string): Promise<void> {
      this.durationMs(runId);
    }

    async handleToolError(_err: unknown, runId: string): Promise<void> {
      this.starts.delete(runId);
    }

    // ── Helpers ────────────────────────────────────────

    private durationMs(runId: string): number {
      const start = this.starts.get(runId);
      this.starts.delete(runId);
      if (start === undefined) {
        return 0;
      }
      return Date.now() - start;
    }
  }

  return new NavaiaForgeCallbackImpl();
}

// ── Pure helpers (testable without @langchain/core) ────────────

/**
 * Pull input/output tokens and cost out of an LLMResult-like object.
 *
 * LangChain's response shape varies across versions and providers. The
 * canonical place is `output.llmOutput.tokenUsage`, with `usage_metadata`
 * on individual messages a newer alternative.
 */
export function extractTokenUsage(response: unknown): UsageNumbers {
  const out: UsageNumbers = { input: 0, output: 0, costUsd: 0 };
  if (!response || typeof response !== "object") {
    return out;
  }
  const r = response as Record<string, unknown>;

  const llmOutput = r["llmOutput"];
  if (llmOutput && typeof llmOutput === "object") {
    const o = llmOutput as Record<string, unknown>;
    const usageRaw = o["tokenUsage"] ?? o["token_usage"] ?? o["usage"];
    if (usageRaw && typeof usageRaw === "object") {
      const u = usageRaw as Record<string, unknown>;
      out.input = toInt(u["inputTokens"] ?? u["input_tokens"] ?? u["promptTokens"] ?? u["prompt_tokens"]);
      out.output = toInt(u["outputTokens"] ?? u["output_tokens"] ?? u["completionTokens"] ?? u["completion_tokens"]);
    }
  }

  if (out.input === 0 && out.output === 0) {
    const generations = r["generations"];
    if (Array.isArray(generations)) {
      for (const batch of generations) {
        if (!Array.isArray(batch)) continue;
        for (const gen of batch) {
          const message = (gen as Record<string, unknown> | undefined)?.[
            "message"
          ];
          const meta = (message as Record<string, unknown> | undefined)?.[
            "usage_metadata"
          ];
          if (meta && typeof meta === "object") {
            const m = meta as Record<string, unknown>;
            out.input = toInt(m["input_tokens"]);
            out.output = toInt(m["output_tokens"]);
            if (out.input || out.output) return out;
          }
        }
      }
    }
  }

  return out;
}

/** Best-effort model name extraction from an LLMResult-like object. */
export function extractModelName(response: unknown): string | null {
  if (!response || typeof response !== "object") {
    return null;
  }
  const r = response as Record<string, unknown>;
  const llmOutput = r["llmOutput"];
  if (llmOutput && typeof llmOutput === "object") {
    const o = llmOutput as Record<string, unknown>;
    const name = o["modelName"] ?? o["model_name"] ?? o["model"];
    if (typeof name === "string" && name.length > 0) {
      return name;
    }
  }
  return null;
}

function toInt(value: unknown): number {
  if (typeof value === "number" && Number.isFinite(value)) {
    return Math.trunc(value);
  }
  if (typeof value === "string") {
    const parsed = Number.parseInt(value, 10);
    return Number.isFinite(parsed) ? parsed : 0;
  }
  return 0;
}
